### PR TITLE
Problem: Missing blockscout getTokenHolders support (fix #149)

### LINKED ISSUE #149: Problem: Missing blockscout getTokenHolders support
Check https://cronos.org/explorer/api-docs#token

### PR DESCRIPTION
**Pull request by elecming**
Close:
- #149

Solution:
Add get_token_holders function

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-com/play-cpp-sdk/blob/main/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest main? 
- [x] Have you checked your code compiles? (`cargo build`)
- [x] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`cargo test`)
- [ ] Have you checked your code formatting is correct? (`cargo fmt -- --check --color=auto`)
- [ ] Have you checked your basic code style is fine? (`cargo clippy`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`cargo audit`)
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-com/play-cpp-sdk/blob/main/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-com/play-cpp-sdk/blob/main/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-com/play-cpp-sdk/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: extra-cpp-bindings/src/lib.rs [hunk 1]
```diff
@@ -175,48 +184,87 @@ mod ffi {
...
-        /// Crypto.com Pay API
+        /// given the BlockScout REST API base url and the contract address (hexadecimal),
+        ///
+        /// page: A nonnegative integer that represents the page number to be used for
+        /// pagination. 'offset' must be provided in conjunction.
+        ///
+        /// offset: A nonnegative integer that represents the maximum number of records to
+        /// return when paginating. 'page' must be provided in conjunction. token ids.
+        ///
+        /// It will return the list of owners and balances (sorting from largetst to smallest), but no
```
Comment (tomtau) [inline extra-cpp-bindings/src/lib.rs]: "but no"?

File: extra-cpp-bindings/src/lib.rs [hunk 2]
```diff
@@ -175,48 +184,87 @@ mod ffi {
...
+        ///
+        /// (ref: https://cronos.org/explorer/api-docs#token)
+        pub fn get_token_holders(
+            blockscout_base_url: String,
+            contract_address: String,
+            page: String,
+            offset: String,
```
Comment (tomtau) [inline extra-cpp-bindings/src/lib.rs]: could page and offset be unsigned integers instead of strings?

**Review by tomtau — APPROVED** [review]
ok, but for page/offset, is it possible to provide integers instead to make it less error-prone?

File: extra-cpp-bindings/src/lib.rs (hunk 1 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: ok...

File: extra-cpp-bindings/src/lib.rs (hunk 2 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: let me check

File: extra-cpp-bindings/src/lib.rs (hunk 1 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: done

File: extra-cpp-bindings/src/lib.rs (hunk 2 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: done

File: extra-cpp-bindings/src/lib.rs [hunk 3]
```diff
@@ -132,6 +132,15 @@ mod ffi {
         pub token_type: String,
     }
 
+    /// Token holder detail from BlockScout API
```
Comment (leejw51crypto) [inline extra-cpp-bindings/src/lib.rs:135]: maybe add comment ` tokenid is not supported`, need to use account address to fetch tokenid ?

File: extra-cpp-bindings/src/lib.rs [hunk 4]
```diff
@@ -132,6 +132,15 @@ mod ffi {
...
+    #[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
```
Comment (leejw51crypto) [inline extra-cpp-bindings/src/lib.rs:138]: PartialEq is necessary?

**Review by leejw51crypto — APPROVED** [review]
lgtm

File: extra-cpp-bindings/src/lib.rs (hunk 3 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs:135]: done

File: extra-cpp-bindings/src/lib.rs (hunk 4 above)
Comment (elecming) [reply extra-cpp-bindings/src/lib.rs:138]: yes, just keep the same as other types, such as `RawTokenResult`, and `RawTxDetail`

Comment (tomtau) [reply extra-cpp-bindings/src/lib.rs:138]: I think it's needed for tests / assert equals?

File: extra-cpp-bindings/src/lib.rs [hunk 5]
```diff
@@ -659,4 +740,15 @@ mod test {
...
+
+    #[test]
+    pub fn test_get_token_holders() {
+        let actual_result = get_token_holders(
+            "https://blockscout.com/xdai/mainnet/api",
+            "0xed1efc6efceaab9f6d609fec89c9e675bf1efb0a",
+            1,
+            10,
+        );
+        println!("{:?}", actual_result);
```
Comment (tomtau) [inline extra-cpp-bindings/src/lib.rs]: assert some properties about the expected result content?

Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: ok

Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: The result (address and the balance) is likely to be changed in the future. So it is difficult to assert the content.

Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: I'll still add it. Update in the future if necessary .

Comment (elecming) [reply extra-cpp-bindings/src/lib.rs]: done

**Pull request merged**